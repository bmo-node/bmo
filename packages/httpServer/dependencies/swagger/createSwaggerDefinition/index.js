import { has, each, set, get } from 'lodash'
import joi2Swagger from 'joi-to-swagger'
import httpStatus from 'http-status-codes'

const PATH_DELIMITER = '/'
const PARAM_DELIMITER = ':'
const REQUIRED = 'required'
const SCHEMA_TYPES = {
  QUERY_PARAMS: 'queryParams',
  RESPONSE_BODY: 'responseBody',
  REQUEST_BODY: 'requestBody',
  PATH_PARAMS: 'pathParams'
}
const { REQUEST_BODY, RESPONSE_BODY, QUERY_PARAMS, PATH_PARAMS } = SCHEMA_TYPES
const {
  OK,
  CREATED,
  BAD_REQUEST,
  INTERNAL_SERVER_ERROR,
  UNAUTHORIZED,
  NO_CONTENT,
  ACCEPTED
} = httpStatus

const OPEN_API_VERSION = '3.0.0'
export default () => (routes, { title, description, contact, version }) => {
  const components = getComponents(routes, '', {})
  const paths = getPaths(routes, components)
  return {
    openapi: OPEN_API_VERSION,
    info: {
      title,
      description,
      contact,
      version
    },
    paths,
    components
  }
}

const definePathParam = ({ name, type, flags }) => {
  const { description, ...metadata } = flags
  const detailedDescription = description ?
    `${description}  ---  Additional Parameter Context: ${JSON.stringify(metadata)}` :
    `${name} path param`
  return {
    name,
    in: 'path',
    description: detailedDescription,
    required: true,
    style: 'simple',
    schema: { type }
  }
}

const getPathParams = ({ path, schema = {}}) => {
  const parts = path.split(PATH_DELIMITER)
  const pathSchema = has(schema, `${PATH_PARAMS}`) ?
    schema[PATH_PARAMS].describe().keys :
    {}
  return parts.filter(part => part[0] === PARAM_DELIMITER)
    .map(part => {
      const name = part.substring(1)
      return definePathParam({
        name, type: 'string', flags: {}, ...pathSchema[name]
      })
    })
}

const formatPathParams = path => {
  return path
    .split(PATH_DELIMITER)
    .map(part => part[0] === PARAM_DELIMITER ? `{${part.substring(1)}}` : part)
    .join(PATH_DELIMITER)
}

const defineQueryParam = ({ name, flags, type }) => {
  const { description, presence, ...metadata } = flags
  const detailedDescription = description ?
    `${description}  ---  Additional Parameter Context: ${JSON.stringify(metadata)}` :
    `${name} path param`
  return {
    name,
    in: 'query',
    description: detailedDescription,
    required: presence === REQUIRED,
    style: 'simple',
    schema: { type }
  }
}

const getQueryParams = route => {
  const queryParams = has(route.schema, QUERY_PARAMS) ?
    route.schema[QUERY_PARAMS].describe().keys :
    {}
  return Object.entries(queryParams)
    .map(([ key, value ]) => defineQueryParam({
      name: key,
      flags: {},
      type: 'string',
      ...value
    }))
}

const jsonSchema = (schema, description) => ({
  description,
  content: {
    'application/json': {
      schema: {
        $ref: `#/components/schemas/${schema}`
      }
    }
  }
})

const requestBody = schema => `${schema}-${REQUEST_BODY}`

const responseBody = schema => `${schema}-${RESPONSE_BODY}`
const statusDescriptions = {
  [OK]: 'OK',
  [CREATED]: 'Created',
  [ACCEPTED]: 'Accepted',
  [NO_CONTENT]: 'No Content'
}
const DEFAULT_RESPONSES = {
  [BAD_REQUEST]: jsonSchema(`${responseBody('error')}`, 'Bad Request'),
  [UNAUTHORIZED]: jsonSchema(`${responseBody('error')}`, 'Unauthorized'),
  [INTERNAL_SERVER_ERROR]: jsonSchema(`${responseBody('error')}`, 'Internal Server Error')
}

const defaultHttpResponse = (schemaName, successStatusCode) => ({
  [successStatusCode]: jsonSchema(`${responseBody(schemaName)}`, statusDescriptions[successStatusCode]),
  ...DEFAULT_RESPONSES
})
const successStatus = 'successStatus'
const formatters = {
  delete: (schemaName, route = {}) => ({
    responses: {
      ...defaultHttpResponse(schemaName, get(route, successStatus, OK))
    }
  }),
  get: (schemaName, route) => ({
    responses: {
      ...defaultHttpResponse(schemaName, get(route, successStatus, OK))
    }
  }),
  put: (schemaName, route) => ({
    requestBody: jsonSchema(`${requestBody(schemaName)}`),
    responses: {
      ...defaultHttpResponse(schemaName, get(route, successStatus, OK))
    }
  }),
  patch: (schemaName, route) => ({
    requestBody: jsonSchema(`${requestBody(schemaName)}`),
    responses: {
      ...defaultHttpResponse(schemaName, get(route, successStatus, OK))
    }
  }),
  post: (schemaName, route) => ({
    requestBody: jsonSchema(`${requestBody(schemaName)}`),
    responses: {
      ...defaultHttpResponse(schemaName, get(route, successStatus, CREATED))
    }
  })
}

const formatRequestParams = (schemaName, httpMethod, route) => {
  if (formatters[httpMethod]) {
    return formatters[httpMethod](schemaName, route)
  }
}

const getPathDefinition = route => {
  const pathParams = getPathParams(route)
  const queryParams = getQueryParams(route)
  const parameters = pathParams.concat(queryParams)
  const formattedPath = formatPathParams(route.path)
  const schemaName = getComponentName(route)
  const method = route.method.toLowerCase()
  let schemaDef = {}
  if (route.schema) {
    schemaDef = formatRequestParams(schemaName, method, route)
  }

  return {
    key: `${formattedPath}.${method}`,
    value: {
      summary: `${schemaName}`,
      description: route.description,
      parameters,
      ...schemaDef
    }
  }
}

const getPaths = (routes, components) => {
  const paths = {}
  routes.forEach(route => {
    routes[route.path] = routes[route.path] || {}
    const definition = getPathDefinition(route, route.path, components)
    set(paths,
      definition.key,
      definition.value)
  })
  return paths
}

const getComponentName = route => {
  return route.name ? route.name : `${route.method}-${route.path.replace(/\//gi, '-')}`
}

const getComponents = (routes, parentPath, aggregate = {}) => {
  each(routes, route => {
    if (route.schema) {
      const componentName = getComponentName(route)
      each(route.schema, (schema, type) => {
        const { swagger } = joi2Swagger(schema, aggregate)
        set(aggregate, `schemas.${componentName}-${type}`, swagger)
      })
    }
  })

  return aggregate
}
