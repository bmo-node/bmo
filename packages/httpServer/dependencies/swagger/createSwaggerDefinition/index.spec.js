import createSwaggerDefinition from '.'
import joi from 'joi'
const serverInfo = {
  title: 'testServer',
  description: 'test server',
  contact: 'no one',
  version: '0.0.1'
}
const testRoutes = [{
  description: 'a  test path to demo things work',
  method: 'GET',
  schema: {
    requestBody: joi.object().keys({
      test: joi.number().description('test date').default(0).allow(1, 2, 3)
    }),
    responseBody: joi.object().keys({
      test: joi.date().default(new Date())
    }),
    queryParams: joi.object().keys({
      testParam: joi.number().allow(1, 2, 3).description('a test').valid(1, 2, 3).default(0).required()
    }),
    pathParams: joi.object().keys({
      idtest: joi.number().allow(1).description('asdf').valid(1, 2, 3)
    })
  },
  path: '/api/test/v1/:idtest',
  handler: ctx => {
    ctx.body = { res: 1 }
  }
},
{
  description: 'demo POST',
  method: 'POST',
  path: '/api/test/v2/:boom',
  handler: ctx => {
    ctx.body = { res: 1 }
  }
}]
describe('createSwaggerDefinition', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('defined routes should match snapshot', async () => {
    const apiDocs = createSwaggerDefinition()(testRoutes, serverInfo)
    expect(apiDocs).toMatchSnapshot()
  })
})
