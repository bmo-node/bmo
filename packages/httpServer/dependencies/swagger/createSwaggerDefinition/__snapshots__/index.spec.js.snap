// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSwaggerDefinition defined routes should match snapshot 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "GET--api-test-v1-:idtest-pathParams": Object {
        "additionalProperties": false,
        "properties": Object {
          "idtest": Object {
            "description": "asdf",
            "enum": Array [
              1,
              2,
              3,
            ],
            "format": "float",
            "type": "number",
          },
        },
        "type": "object",
      },
      "GET--api-test-v1-:idtest-queryParams": Object {
        "additionalProperties": false,
        "properties": Object {
          "testParam": Object {
            "description": "a test",
            "enum": Array [
              1,
              2,
              3,
            ],
            "format": "float",
            "type": "number",
          },
        },
        "required": Array [
          "testParam",
        ],
        "type": "object",
      },
      "GET--api-test-v1-:idtest-requestBody": Object {
        "additionalProperties": false,
        "properties": Object {
          "test": Object {
            "description": "test date",
            "format": "float",
            "type": "number",
          },
        },
        "type": "object",
      },
      "GET--api-test-v1-:idtest-responseBody": Object {
        "additionalProperties": false,
        "properties": Object {
          "test": Object {
            "default": 2023-01-17T22:30:33.174Z,
            "format": "date-time",
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": "no one",
    "description": "test server",
    "title": "testServer",
    "version": "0.0.1",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/api/test/v1/{idtest}": Object {
      "get": Object {
        "description": "a  test path to demo things work",
        "parameters": Array [
          Object {
            "description": "asdf  ---  Additional Parameter Context: {\\"only\\":true}",
            "in": "path",
            "name": "idtest",
            "required": true,
            "schema": Object {
              "type": "number",
            },
            "style": "simple",
          },
          Object {
            "description": "a test  ---  Additional Parameter Context: {\\"only\\":true,\\"default\\":0}",
            "in": "query",
            "name": "testParam",
            "required": true,
            "schema": Object {
              "type": "number",
            },
            "style": "simple",
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/GET--api-test-v1-:idtest-responseBody",
                },
              },
            },
            "description": "OK",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/error-responseBody",
                },
              },
            },
            "description": "Bad Request",
          },
          "401": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/error-responseBody",
                },
              },
            },
            "description": "Unauthorized",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/error-responseBody",
                },
              },
            },
            "description": "Internal Server Error",
          },
        },
        "summary": "GET--api-test-v1-:idtest",
      },
    },
    "/api/test/v2/{boom}": Object {
      "post": Object {
        "description": "demo POST",
        "parameters": Array [
          Object {
            "description": "boom path param",
            "in": "path",
            "name": "boom",
            "required": true,
            "schema": Object {
              "type": "string",
            },
            "style": "simple",
          },
        ],
        "summary": "POST--api-test-v2-:boom",
      },
    },
  },
}
`;
